crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -4184,6 +4193,117 @@ impl<'db> TypeInferenceBuilder<'db> {
         Ok(builder.build())
     }
 
+    fn infer_type_api_special_form(
+        &mut self,
+        special_form: TypeApiSpecialForm,
+        arguments: &ast::Expr,
+    ) -> Type<'db> {
+        let db = self.db();
+
+        let argument_types = match (special_form, arguments) {
+            (_, ast::Expr::Tuple(tuple)) => Either::Left(
+                tuple
+                    .iter()
+                    .map(|element| self.infer_type_expression(element)),
+            ),
+            (TypeApiSpecialForm::TypeOf, expr) => {
+                Either::Right(std::iter::once(self.infer_expression(expr)))
+            }
+            (_, expr) => Either::Right(std::iter::once(self.infer_type_expression(expr))),
+        };
+
+        type_api::resolve_special_form(db, special_form, argument_types).unwrap_or_else(
+            |TypeApiArgumentsError { expected, actual }| {
+                self.context.report_lint(
+                    &TYPE_API_WRONG_ARITY,
+                    arguments.into(),
+                    format_args!(
+                        "Expected {expected} type argument{}, got {actual}",
+                        if expected == 1 { "" } else { "s" },
+                    ),
+                );
+                Type::Unknown
+            },
+        )
+    }
+
+    fn try_infer_type_api_predicate(
+        &mut self,
+        function: Type<'db>,
+        arguments: &ast::Arguments,
+    ) -> Option<Type<'db>> {
+        let db = self.db();
+
+        match function
+            .into_function_literal()
+            .and_then(|function| function.known(db))
+            .and_then(KnownFunction::into_type_api_function)
+        {
+            Some(api_function) => {
+                let argument_types = arguments.args.iter().map(|arg| {
+                    if api_function.is_static_assert() {
+                        self.infer_expression(arg)
+                    } else {
+                        self.infer_type_expression(arg)
+                    }
+                });
+
+                let result = type_api::resolve_predicate(db, api_function, argument_types);
+
+                match result {
+                    Ok(ty) => Some(ty),
+                    Err(TypeApiPredicateError::ArgumentsError(TypeApiArgumentsError {
+                        expected,
+                        actual,
+                    })) => {
+                        self.context.report_lint(
+                            &TYPE_API_WRONG_ARITY,
+                            arguments.into(),
+                            format_args!(
+                                "Expected {expected} argument{}, got {actual}",
+                                if expected == 1 { "" } else { "s" },
+                            ),
+                        );
```

Comment:
I would expect this to emit the same diagnostic as any other wrong-arity call, not a special diagnostic specific to red-knot type-api.

Reply:
Ok, rebased my branch on top of your call-signature checking branch, so this now uses your implementation.